stepper_eip_driver/src/stepper_node.cpp

```diff
@@ -95,6 +97,9 @@ int main(int argc, char *argv[])
   ros::Publisher stepper_pub = nh.advertise<stepper_inputs>("stepper_inputs", 1);
   ros::Publisher status_pub = nh.advertise<stepper_status>("stepper_status", 1);
 
+  sensor_msgs::JointState joint_state;
```

Comment:
If this is just a time_stamped position, can we just call it that? As I commented elsewhere, I think we should endeavor to keep things application agnostic.

And I don't really see the use case here. I mean, the current position is already available; the name can come along with the topic; and I think ROS msg's have built in time stamp info. So I think some subscriber node should be able  to get everything it need's from the existing current_position topic. What am I missing?

Reply:
The underlying motivation is to broadcast the state of the stepper axis as a tf transform, which makes it easier to chain together several axes (e.g. in a gantry) and find the cumulative transform along their kinematic chain. The various ROS-I robot packages do this by launching a `robot_state_publisher` node, defining the relationships between the axes in the robot in a URDF, and publishing the state of each motion axis to the `/joint_states` topic as `sensor_msgs::JointState` messages. The `robot_state_publisher` does all the work behind the scenes to parse the URDF and generate the tf messages, which is a lot cleaner than manually setting up tf publishers.

The missing piece was publishing the data in the `stepper_status` message as a `sensor_msgs::JointState` message. One option was creating my own node that would subscribe to the `stepper_status` topic for each motor, convert the stepper axis positions to `JointState` messages, and publish them to `/joint_states`. I decided that it would be better to include this functionality in `stepper_node`, since the `robot_state_publisher` is a relatively common tool and I didn't want to have to recreate a utility converter node on future projects. It's still the responsibility of the user to model the stepper axis correctly in their URDF, set up the `robot_state_publisher`, and give each axis a unique name -- the last two things can happen in the same launch file.

[The dynamixel-workbench package takes a very similar approach](https://github.com/ROBOTIS-GIT/dynamixel-workbench/blob/529aaded82fa1876779791fab3705d2606f7028a/dynamixel_workbench_controllers/src/dynamixel_workbench_controllers.cpp#L457-L490) and publishes its servo states as both generic `sensor_msgs::JointState` messages and custom `dynamixel_workbench_msgs::DynamixelStateList` messages. They also provide a flag to toggle the creation of the JointState publisher, which is probably a useful thing to add here.

I guess the tl;dr is that publishing `JointState` messages is an application-agnostic way to communicate the state of the motor that meshes nicely with other ROS capabilities.

Reply:
I think it's good idea idea to be looking at how more mature projects work; I hadn't considered that. Thanks! So let's use something like their _use_joint_states_topic_ parameter to turn it on or off. Consider the application that wouldn't be using this in a manipulator. (I once used a stepper motor to drive a volumetric pump.) I think a goal should be to have some very minimal mode of operation as a default, where the base mfg interface is the only thing exposed, and turn on more advanced features as required.

Reply:
In the end this is I believe an example of the 'transparent bridge' vs '(opaque) service proxy' patterns, where @wpmccormick's suggestion ("minimal mfg interface") is the former, while @schornakj's work goes into the direction of the latter.

A transparent bridge will import implementation details of beyond the bridge into every application that uses it, making things like conversion of data from devices on the other side the responsibility of consumers of that data (in this case: stepper position/velocity/etc).

A service proxy provides an almost opaque ROS API that completely shields the ROS consumers/clients from whatever is on the other side -- essentially pretends to be exposing the functionality implemented by devices beyond the proxy (node) itself. Conversion to and from units required by ROS would be one of the responsibilities of such a proxy, but it could go as far as coordinating multiple devices (both in time and space) to appear as a single entity to ROS applications (an example here would be a `ros_control` hw interface that provides access to actuators as simple `double` registers, but that actually require much more state and control to properly function (the dynamixels @schornakj referred to are a simple example of something like that)).

---

Most ROS "drivers" I've seen are variants of opaque proxies, but it's not necessarily clear which approach is better. Personally I always like to isolate domains as much as possible though (tend to prefer opaque proxies): it decouples applications and nodes to a greater extend, making replacement and testing easier.

A hybrid approach I've seen is something like what `velodyne_driver` does: the base nodelet essentially implements a transparent bridge, which transforms incoming scans into ROS msgs but does not convert anything. A second nodelet is then used to do the conversion into standard ROS datatypes such as `LaserScan` and `PointCloud2`.

Such a scheme allows users that would want access to the raw data (the "minimal mfg interface") the possibility to get just that, but would still support setting up an opaque service proxy.

Reply:
@wpmccormick I added a boolean parameter  that toggles the joint_state publisher -- basically how it's handled in the dynamixel package.